### ISSUE TITLE
miniupnpd: duplicate rules with nft backend

### ISSUE THREAD

**Issue by stintel** [miniupnp/miniupnp#582]
It seems miniupnpd inserts the same rules again whenever a client requests a mapping:

```
table inet miniupnpd {
        chain forward {
                type filter hook forward priority -25; policy accept;
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
        }
}
table ip miniupnpd {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
        }

        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
        }
}
```

This is with miniupnd-nftables 2.2.3 on OpenWrt, a WIP from openwrt/packages#17094. Other versions:
* libmnl 1.0.4
* libnftnl 1.2.0
* nftables 1.0.0

**Comment by stintel** [comment]
Aside from that, I'm sporadically seeing this in the logs:

```
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:17:11 2021 daemon.err miniupnpd[2988]: nft_send_rule(0x3fff80ab3820, 6, 2) send_batch failed -4
Sat Nov 27 01:23:25 2021 daemon.notice miniupnpd[2988]: shutting down MiniUPnPd
Sat Nov 27 01:23:25 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:23:25 2021 daemon.err miniupnpd[2988]: chain_op(5, 1, miniupnpd, forward, filter, 2, -25) send_batch failed -4
Sat Nov 27 01:23:25 2021 daemon.err miniupnpd[2988]: send_batch: mnl_cb_run returned -1
Sat Nov 27 01:23:25 2021 daemon.err miniupnpd[2988]: chain_op(5, 2, miniupnpd, prerouting, nat, 0, -100) send_batch failed -4
```

I noticed that earlier today when my Xbox failed to get open NAT via UPnP. Not sure if the two are related, if you want me to open a second issue for it, let me know and I will do so.

**Comment by attila-lendvai** [comment]
i'm also seeing a flood of these regularly on openwrt 22.03.0-rc4:

```
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77d619e0, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 54351 TCP because it has expired
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77d617f0, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 33295 TCP because it has expired
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77d61600, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 54351 TCP because it has expired
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77d61410, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 33295 TCP because it has expired
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77eb0610, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 33295 TCP because it has expired
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: send_batch: mnl_cb_run returned -1
Thu Jun 30 13:38:48 2022 daemon.err miniupnpd[3840]: nft_send_rule(0x77eb0040, 8, 2) send_batch failed -4
Thu Jun 30 13:38:48 2022 daemon.notice miniupnpd[3840]: remove port mapping 54116 TCP because it has expired
```

**Comment by yujincheng08** [comment]
@stintel @attila-lendvai I think this is because tables of miniupnpd are removed. I have no idea why they got removed (at least I manually restart fw4 those tables are present), but in my devices, when those tables were removed, I saw the err `send_batch failed -4` whenever I try to add a UPnP rule. So I think they are irrelevant.

**Comment by miniupnp** [comment]
@stintel is this issue still relevant ?
Do you reproduce it with miniupnpd 2.3.0 ?

**Comment by axet** [comment]
```
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 51414 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 45573 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 45574 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50540, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 51415 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50b50, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 45575 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdfd70, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 29577 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x555d4860, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 45576 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 51416 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:03:32 2022 daemon.notice miniupnpd[3315]: remove port mapping 45577 TCP because it has expired
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:03:32 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 29577 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 29577 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45571 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50540, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdfd70, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45572 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50b50, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x555d4860, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 29577 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 51414 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45573 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50540, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45574 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdfd70, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50b50, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 51415 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x555d4860, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45575 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 29577 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45576 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50540, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdfd70, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 51416 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77d50b50, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45577 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x555d4860, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 45578 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5dc0, 8, 2) send_batch failed -4
Sat Oct 15 12:33:42 2022 daemon.notice miniupnpd[3315]: remove port mapping 30711 TCP because it has expired
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 12:33:42 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cf5bc0, 8, 2) send_batch failed -4
Sat Oct 15 14:53:15 2022 daemon.warn miniupnpd[3315]: HTTP Connection from 10.10.5.92 closed unexpectedly
Sat Oct 15 14:53:15 2022 daemon.warn miniupnpd[3315]: HTTP Connection from 10.10.5.92 closed unexpectedly
Sat Oct 15 14:53:25 2022 daemon.err miniupnpd[3315]: send_batch: mnl_cb_run returned -1
Sat Oct 15 14:53:25 2022 daemon.err miniupnpd[3315]: nft_send_rule(0x77cdf970, 8, 2) send_batch failed -4
Sat Oct 15 14:55:07 2022 authpriv.info dropbear[3440]: Child connection from fc00:10:10:5::e9f:45598
Sat Oct 15 14:55:07 2022 authpriv.notice dropbear[3440]: Pubkey auth succeeded for 'root' with ssh-rsa key SHA256:/WDh9rVXUcintsjaGfzVp09ocX3/9aIqNhisjaUgcDE from fc00:10:10:5::e9f:45598
root@OpenWrt:~# miniupnpd --version
miniupnpd 2.3.0 v22.03.1 Oct  7 2022
using netfilter(nftables) backend

```

**Comment by axet** [comment]
Same for 2.3.1. Rules not removed from 'nft list ruleset'

```
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: send_batch: mnl_cb_run returned -1
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: nft_send_rule(0x77ef8a30, 8, 2) send_batch failed -4
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: send_batch: mnl_cb_run returned -1
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: nft_send_rule(0x77de39d0, 8, 0) send_batch failed -4
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: send_batch: mnl_cb_run returned -1
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: nft_send_rule(0x77de39d0, 8, 2) send_batch failed -4
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: send_batch: mnl_cb_run returned -1
Thu Dec 29 21:00:49 2022 daemon.err miniupnpd[3526]: nft_send_rule(0x55584860, 8, 0) send_batch failed -4
root@OpenWrt:~# miniupnpd --version
miniupnpd 2.3.1 v22.03.2 Oct 14 2022
using netfilter(nftables) backend
```

**Comment by brttbndr** [comment]
I am running OpenWrt 22.03.2 on an EdgeRouter 4 (mips64_octeonplus)

One source of duplicate rules seems to be fixed for me in miniupnpd 2.3.1 (which I built locally as it's not yet available via OpenWrt packages). This is the result of https://github.com/miniupnp/miniupnp/issues/628#issuecomment-1272660846

With 2.3.0 with debug logging on, I was getting:

```
Tue Jan 10 02:15:13 2023 daemon.debug miniupnpd[4596]: port_in_use check port 51668 on nat chain PREROUTING redirected to 10.11.12.129 port 0
Tue Jan 10 02:15:13 2023 daemon.info miniupnpd[4596]: port 51668 protocol udp already in use
Tue Jan 10 02:15:13 2023 daemon.debug miniupnpd[4596]: UPnP permission rule 0 matched : port mapping accepted
Tue Jan 10 02:15:13 2023 daemon.debug miniupnpd[4596]: Check protocol udp for port 51669 on ext_if pppoe-PPP ...
Tue Jan 10 02:15:13 2023 daemon.debug miniupnpd[4596]: timestamp entry added (51669, 17, 6116316)
```

That is -- the existing rule on 51668 is evaluated with `port 0` (b/c of endianess bug) and isn't matched, so a new rule is added on 51669. This repeats and the rules stack up.

I'm not sure how to reconcile my positive result with @axet finding the problem remains in 2.3.1 -- perhaps we have similar symptoms, but a different root cause.

**Comment by miniupnp** [comment]
@brttbndr can you check that my branch [fix-nftnl_rule_get_data](https://github.com/miniupnp/miniupnp/tree/fix-nftnl_rule_get_data) fixes the issue ?

**Comment by brttbndr** [comment]
@miniupnp Confirmed, your branch is a fix.

I built [miniupnpd-nftables_2023-01-14-904dda47-1_mips64_octeonplus.ipk.zip](https://github.com/miniupnp/miniupnp/files/10418861/miniupnpd-nftables_2023-01-14-904dda47-1_mips64_octeonplus.ipk.zip) 
using [OpenWRT single package instructions](https://openwrt.org/docs/guide-developer/toolchain/single.package) and commit 904dda47ed67def29a9cfece592637c6777fc54b
 
[Makefile.patch](https://github.com/miniupnp/miniupnp/files/10418848/Makefile.patch)

With that build installed, I get expected log output like:

```
Jan 14 21:21:06 router miniupnpd[32307]: port_in_use check port 51668 on nat chain PREROUTING redirected to 10.11.12.133 port 51668
Jan 14 21:21:06 router miniupnpd[32307]: timestamp entry found (51668, 17, 329540)
Jan 14 21:21:06 router miniupnpd[32307]: port 51668 udp already redirected to 10.11.12.133:51668, replacing
Jan 14 21:21:06 router miniupnpd[32307]: timestamp entry added (51668, 17, 329840)
Jan 14 21:21:06 router miniupnpd[32307]: table_cb(0xfffbeef810, 0xfffbef0820) fw4 upnp_forward 2
Jan 14 21:21:06 router miniupnpd[32307]: table_cb(0xfffbeefa0c, 0xfffbef0820) fw4 upnp_forward 2
Jan 14 21:21:06 router miniupnpd[32307]: table_cb(0xfffbeefc10, 0xfffbef0820) fw4 upnp_forward 2
Jan 14 21:21:06 router miniupnpd[32307]: table_cb(0xfffbeefe1c, 0xfffbef0820) fw4 upnp_forward 2
Jan 14 21:21:06 router miniupnpd[32307]: table_cb(0xfffbef0020, 0xfffbef0820) fw4 upnp_forward 2
```

and output from `nft list ruleset` shows only unique rules, not an accumulation of duplicates.

**Comment by axet** [comment]
904dda47ed67def29a9cfece592637c6777fc54b works! No log spamming (send_batch failed -4), no duplicate rules, rules removed after lease removed. Everything looks fine!

**Comment by miniupnp** [comment]
fixed version is [miniupnpd-2.3.2.tar.gz](http://miniupnp.free.fr/files/miniupnpd-2.3.2.tar.gz)
